### PR TITLE
Added trace alias to jnp.linalg

### PR DESCRIPTION
**Pull request by vfdev-5**
Related to #21088

cc @Micky774

### CODE REVIEW

**Comment by vfdev-5** [comment]
> Instead of adding an alias in the linalg source file, could we just add a direct import of the source of `jnp.trace` into the linalg namespace file?

I have no strong opinion on the approach. Yes, we can do that as well. In the current approach we sort of hidden that detail into to the `_src`.

File: jax/_src/numpy/linalg.py [hunk 1]
```diff
@@ -1599,7 +1599,7 @@ def matrix_transpose(x: ArrayLike, /) -> Array:
 
 def vector_norm(x: ArrayLike, /, *, axis: int | None = None, keepdims: bool = False,
                 ord: int | str = 2) -> Array:
-  """Computes the vector norm of a vector or batch of vectors.
+  """Compute the vector norm of a vector or batch of vectors.
```
Comment (vfdev-5) [inline jax/_src/numpy/linalg.py:1602]: Drive-by fix according to https://google.github.io/styleguide/pyguide.html 

> The docstring may be descriptive-style ("""Fetches rows from a Bigtable.""") or imperative-style ("""Fetch rows from a Bigtable."""), but the style should be consistent within a file.

File: tests/linalg_test.py [hunk 2]
```diff
@@ -1305,6 +1305,18 @@ def testDiagonal(self, shape, dtype, offset):
     self._CheckAgainstNumpy(np_fun, lax_fun, args_maker)
     self._CompileAndCheck(lax_fun, args_maker)
 
+  def testTrace(self):
```
Comment (vfdev-5) [inline tests/linalg_test.py:1308]: Here I added a simple test instead of running a large set of test case for example compared to above testDiagonal test. The reason is that `jnp.linalg.trace` is a thin wrapper over jnp.trace which should be already extensively tested, so IMO there is no need to run large set of test cases.

File: jax/_src/numpy/linalg.py [hunk 3]
```diff
@@ -2136,3 +2136,49 @@ def cond(x: ArrayLike, p=None):
...
+  """Compute the trace of a matrix.
+
+  JAX implementation of :func:`numpy.linalg.trace`.
+
+  Args:
+    x: array of shape ``(..., M, N)`` and whose innermost two
+      dimensions form MxN matrices for which to take the trace.
+    offset: positive or negative offset from the main diagonal
+      (default: 0).
+    dtype: data type of the returned array (default: None).
```
Comment (jakevdp) [inline jax/_src/numpy/linalg.py:2152]: It would be useful to specify what `None` means in this context.

File: jax/_src/numpy/linalg.py [hunk 4]
```diff
@@ -2136,3 +2136,49 @@ def cond(x: ArrayLike, p=None):
...
+
+  Returns:
+    array containing the traces and whose shape is determined by
+      removing the last two dimensions and storing the traces in the last
```
Comment (jakevdp) [inline jax/_src/numpy/linalg.py:2156]: All lines in the `Returns:` block should have the same indentation

File: jax/experimental/array_api/linalg.py [hunk 5]
```diff
@@ -35,7 +35,6 @@
   vector_norm as vector_norm,
 )
 
-# TODO(micky774): Add trace to jax.numpy.linalg
 from jax.numpy import trace as trace
```
Comment (jakevdp) [inline jax/experimental/array_api/linalg.py:38]: Import from `jax.numpy.linalg`

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/numpy/linalg.py (hunk 3 above)
Comment (vfdev-5) [reply jax/_src/numpy/linalg.py]: I updated the docstring, let me know if it looks good to you

File: jax/_src/numpy/linalg.py [hunk 6]
```diff
@@ -2136,3 +2136,52 @@ def cond(x: ArrayLike, p=None):
...
+      dimensions form MxN matrices for which to take the trace.
+    offset: positive or negative offset from the main diagonal
+      (default: 0).
+    dtype: data type of the returned array (default: ``None``). If ``None``,
+      the returned array will have the same data type as ``x``. If ``x`` is of
+      integer type of precision less than the default integer precision,
+      then the default integer precision is used.
+
+  Returns:
+    array containing the traces and whose shape is determined by
```
Comment (jakevdp) [inline jax/_src/numpy/linalg.py:2158]: I think it would be simpler to say something like `array of batched traces, shape = x.shape[:-2]`

File: jax/_src/numpy/linalg.py [hunk 7]
```diff
@@ -2136,3 +2136,52 @@ def cond(x: ArrayLike, p=None):
...
+
+  JAX implementation of :func:`numpy.linalg.trace`.
+
+  Args:
+    x: array of shape ``(..., M, N)`` and whose innermost two
+      dimensions form MxN matrices for which to take the trace.
+    offset: positive or negative offset from the main diagonal
+      (default: 0).
+    dtype: data type of the returned array (default: ``None``). If ``None``,
+      the returned array will have the same data type as ``x``. If ``x`` is of
```
Comment (jakevdp) [inline jax/_src/numpy/linalg.py:2153]: Shorter: ```If ``None``, then output dtype will match the dtype of ``x``, promoted to default precision in the case of integer types.```

**Review by jakevdp — CHANGES_REQUESTED** [review]
Sorry, just realized that this new function is missing from the docs here: https://github.com/google/jax/blob/main/docs/jax.scipy.rst#jaxscipylinalg

**Comment by vfdev-5** [comment]
> Sorry, just realized that this new function is missing from the docs here: https://github.com/google/jax/blob/main/docs/jax.scipy.rst#jaxscipylinalg

I should be missing something as I'm not sure to understand your point here. As far as I understand, there is no `jax.scipy.linalg.trace` function and so no docs for it.

**Comment by jakevdp** [comment]
Sorry, I meant to point to the numpy docs: https://github.com/google/jax/blob/main/docs/jax.numpy.rst#jaxnumpylinalg

**Review by jakevdp — APPROVED** [review]
